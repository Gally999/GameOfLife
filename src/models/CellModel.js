import PropTypes from 'prop-types';

import { Alive, Dead } from './CellState';

class CellModel {
  static get shape() {
    return PropTypes.shape({
      id: PropTypes.string.isRequired,
      status: PropTypes.oneOfType([
        PropTypes.instanceOf(Alive),
        PropTypes.instanceOf(Dead),
      ]).isRequired,
    });
  }

  constructor(id, status) {
    this.id = id;
    this.status = status;
  }

  withState(state) {
    return new CellModel(this.id, state);
  }

  get isAlive() {
    return this.status instanceof Alive;
  }

  get isDead() {
    return this.status instanceof Dead;
  }

  evolvesFrom(liveNeighbors) {
    if (this.isDead && liveNeighbors === 3) {
      return this.withState(new Alive());
    }
    if (this.isAlive && liveNeighbors !== 2 && liveNeighbors !== 3) {
      return this.withState(new Dead());
    }
    return this;
  }
}

export default CellModel;
