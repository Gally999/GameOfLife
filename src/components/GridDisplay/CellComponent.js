import React, { Component } from 'react';

import './CellComponent.css';
import CellModel from '../../models/CellModel';

class CellComponent extends Component {
  constructor(props) {
    super(props);
    this.switchState = this.switchState.bind(this);
  }

  switchState() {
    const { cell, onCellClicked } = this.props;
    onCellClicked(cell);
  }

  render() {
    const { cell } = this.props;
    return (
      <td
        className={cell.isAlive ? 'alive' : 'dead'} 
        onClick={this.switchState} 
      ></td>
    );
  }
}

CellComponent.propTypes = {
  cell: CellModel.shape.isRequired,
}

export default CellComponent;
