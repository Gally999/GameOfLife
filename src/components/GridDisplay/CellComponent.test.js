import React from 'react';

import CellComponent from './CellComponent';
import CellModel from '../../models/CellModel';

describe('CellComponent', () => {
  it('renders the cell component correctly', () => {
    const subject = (
      <CellComponent
        cell={CellModel.empty()}
        onCellClicked={() => {}}
      />
    );
    expect(subject).toMatchSnapshot();
  });
});
