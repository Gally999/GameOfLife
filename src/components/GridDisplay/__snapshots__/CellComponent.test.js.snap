// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CellComponent renders the cell component correctly 1`] = `
<CellComponent
  cell={
    Cell {
      "id": "",
      "status": Dead {},
    }
  }
  onCellClicked={[Function]}
/>
`;
